add KEP 2420: Reducing Kubernetes Build Maintenance (#2421)
kubernetes/enhancements: keps/sig-testing/2420-reducing-kubernetes-build-maintenance/README.md

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### Test Plan

Non-blocking “make build” equivalents to kubernetes/kubernetes “bazel” CI jobs will be introduced (if they don’t already exist).
When the new jobs provide equivalent signal, they will be moved to blocking, and the old jobs will be retired.

This is relevant for at least the following jobs:
 - `pull-kubernetes-bazel-test` (this can be converted to ~ `make test`)

>  - `pull-kubernetes-bazel-test` (this largely overlaps with other presubmits, if not for testing ~`bazel build //...` and can  likely be removed)

Reviewer: duplicate lines?

Author: this one should say `pull-kubernetes-bazel-build`
```suggestion
 - `pull-kubernetes-bazel-test` (this can be converted to ~ `make test`)
 - `pull-kubernetes-bazel-build` (this largely overlaps with other presubmits, if not for testing ~`bazel build //...` and can  likely be removed)
```

Author: this is fixed.

#### Alpha -> Beta Graduation

- Gather feedback from developers and surveys
- Complete features A, B, C
- Tests are in Testgrid and linked in KEP

#### Beta -> GA Graduation

- N examples of real-world usage
- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

#### Removing a Deprecated Flag

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag

**For non-optional features moving to GA, the graduation criteria must include 
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
-->

This will be declared stable/GA when:
- All kubernetes/kubernetes CI jobs use the preexisting make build system

> - Bazel-related source files and related tooling are removed from the kubernetes/kubernetes repository on currently supported release branches and the current development branch

Reviewer: is the plan to adjust `master` and then let the bazel-ed release branches age out, or to actively remove bazel from release branches? when we made other rippling changes like switching from godeps to gomod for vendor management, we let release branches age out rather than try to pick back.

Author: Good question: I think we should let it age out. Personally I think we should avoid changes that are not bug-fixes in these branches in general.

Author: I will rewrite this section to reflect that.

Author: I've added a commit reflecting this.

Author: it will cut both ways, cherry picks will include (or exclude and need them to be added) BUILD changes in some cases (when. source files are added / removed, dependencies change).

there is some simplicity in migrating all branches in this case, but I'm generally against manipulating CI / sources for existing branches, except for bug fixes.

Author: discussed further: we'll go forward with focus on the devel branch, planning to leave the release branches, and revisit if somehow it turns out less practical than expected.

- Simplify Golang upgrades for SIG Release
- Remove qualification gaps caused by duplicate-but-slightly-different binary builds
- Remove testing gaps caused by duplicate-but-slightly-different test-invocation methods
- Empower broader community maintenance of tests by converging on Golang testing standards

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->
- Support importing github.com/kubernetes/kubernetes as a library
  - This has never been supported, and is orthogonal to the decision of maintaining duplicate build systems
  - Code under kubernetes/kubernetes/src/staging/k8s.io/foo is intended to be imported as k8s.io/foo from the staged copies at github.com/kubernetes/foo; anything else is internal-only and not supported for import, bazel or otherwise.
- Removing Bazel from sub-projects other than the core repo
  - Bazel comes with tradeoffs, each subproject can make its own decision in this regard. This KEP **only** covers the main Kubernetes repository, and nothing else.
- Improving the previously existing make build

>   - We should strongly consider improving the implementation and behavior of this build in the future, but this is largely orthogonal to whether we should consider maintaining two build systems.

Reviewer 2: ```suggestion
  - We should strongly consider improving the implementation and behavior of this build in the future, but this is largely orthogonal to whether we should consider maintaining two build systems.  An anticipated outcome of this KEP is increased bandwidth available to improve our single build system.
```

Author: This is added.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

> For the development (main/master) branch only, **NOT** the existing release branches:

Reviewer 3: 👍🏾

Author: considered, and rejected for initial implementation, but something we'll keep an eye on.

Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].

#### Alpha -> Beta Graduation

- Gather feedback from developers and surveys
- Complete features A, B, C
- Tests are in Testgrid and linked in KEP

#### Beta -> GA Graduation

- N examples of real-world usage
- N installs
- More rigorous forms of testing—e.g., downgrade tests and scalability tests
- Allowing time for feedback

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

#### Removing a Deprecated Flag

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag

**For non-optional features moving to GA, the graduation criteria must include 
[conformance tests].**

[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
-->

> This will be declared stable/GA when:

Reviewer 3: i like the approach outlined here for GA!

Author: @liggitt @dims, I think this implies then that stable is 1.23 release / 1.24 milestone (when we phase out 1.20 from support). not sure what to call "beta".

Reviewer: very little of the standard KEP template applies to this proposal. I'd probably go more in terms of phase 1, phase 2, done, etc.

Reviewer 3: "beta" would be a bazel-less stable branch? :)

- All kubernetes/kubernetes `master` branch CI jobs use the preexisting make build system
<!-- TODO: as old jobs are rotated out each release anyhow, it may be acceptable to reduce the stable target to just the current development branch -->
- Bazel-related source files and related tooling are removed from the kubernetes/kubernetes repository on currently supported release branches and the current development branch
  - This will only happen for release branches as we phase out support for older releases, rotating in new supported releases that never contained the Bazel build
- Bazel-related configuration/presets are removed from kubernetes/kubernetes CI jobs in kubernetes/test-infra

> As bazel-built artifacts are not built or distributed as part of a kubernetes/kubernetes release, there is no deprecation window required.

Reviewer 2: I wrote the "no deprecation window" part, but in retrospect I want to be clear that this merits a visible announcement.  It's not clear to me that "deprecation" or "action required" are a fit here.
```suggestion
Removal of bazel-related source files from kubernetes/kubernetes will be announced visibly (e.g. in the CHANGELOG, notification to kubernetes-dev@). However, as bazel-built artifacts are not built or distributed as part of a kubernetes/kubernetes release, there is no deprecation window required.
```

Reviewer 2: I'm OK addressing this in followup, sounds like @justaugustus has thoughts on improving how we would announce something like this (ref: https://github.com/kubernetes/community/issues/5344)

**TODO**: This entire section seems completely irrelevant for KEPs that do not
target changes to release artifacts. Delete this section?

<!--

Production readiness reviews are intended to ensure that features merging into
Kubernetes are observable, scalable and supportable; can be safely operated in
production environments, and can be disabled or rolled back in the event they
cause increased failures in production. See more in the PRR KEP at
https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.

The production readiness review questionnaire must be completed and approved
for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.

-->

### Feature Enablement and Rollback

> _This section must be completed when targeting alpha to a release._
> 
> * **How can this feature be enabled / disabled in a live cluster?**
>   - [ ] Feature gate (also fill in values in `kep.yaml`)
>     - Feature gate name:
>     - Components depending on the feature gate:
>   - [ ] Other
>     - Describe the mechanism:
>     - Will enabling / disabling the feature require downtime of the control
>       plane?
>     - Will enabling / disabling the feature require downtime or reprovisioning
>       of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
> 
> * **Does enabling the feature change any default behavior?**
>   Any change of default behavior may be surprising to users or break existing
>   automations, so be extremely careful here.
> 
> * **Can the feature be disabled once it has been enabled (i.e. can we roll back
>   the enablement)?**
>   Also set `disable-supported` to `true` or `false` in `kep.yaml`.
>   Describe the consequences on existing workloads (e.g., if this is a runtime
>   feature, can it break the existing applications?).
> 
> * **What happens if we reenable the feature if it was previously rolled back?**
> 
> * **Are there any tests for feature enablement/disablement?**
>   The e2e framework does not currently support enabling or disabling feature
>   gates. However, unit tests in each component dealing with managing data, created
>   with and without the feature, are necessary. At the very least, think about
>   conversion tests if API types are being modified.

Reviewer 2: ```suggestion
n/a
```

### Rollout, Upgrade and Rollback Planning

> _This section must be completed when targeting beta graduation to a release._
> 
> * **How can a rollout fail? Can it impact already running workloads?**
>   Try to be as paranoid as possible - e.g., what if some components will restart
>    mid-rollout?
> 
> * **What specific metrics should inform a rollback?**
> 
> * **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
>   Describe manual testing that was done and the outcomes.
>   Longer term, we may want to require automated upgrade/rollback tests, but we
>   are missing a bunch of machinery and tooling and can't do that now.
> 
> * **Is the rollout accompanied by any deprecations and/or removals of features, APIs, 
> fields of API types, flags, etc.?**
>   Even if applying deprecation policies, they may still surprise some users.

Reviewer 2: ```suggestion
n/a
```

  checking if there are objects with field X set) may be a last resort. Avoid
  logs or events for this purpose.

* **What are the SLIs (Service Level Indicators) an operator can use to determine 
the health of the service?**
  - [ ] Metrics
    - Metric name:
    - [Optional] Aggregation method:
    - Components exposing the metric:
  - [ ] Other (treat as last resort)
    - Details:

* **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
  At a high level, this usually will be in the form of "high percentile of SLI
  per day <= X". It's impossible to provide comprehensive guidance, but at the very
  high level (needs more precise definitions) those may be things like:
  - per-day percentage of API calls finishing with 5XX errors <= 1%
  - 99% percentile over day of absolute value from (job creation time minus expected
    job creation time) for cron job <= 10%
  - 99,9% of /health requests per day finish with 200 code

* **Are there any missing metrics that would be useful to have to improve observability 
of this feature?**
  Describe the metrics themselves and the reasons why they weren't added (e.g., cost,

>   implementation difficulties, etc.).

Reviewer 2: n/a

### Dependencies

> _This section must be completed when targeting beta graduation to a release._
> 
> * **Does this feature depend on any specific services running in the cluster?**
>   Think about both cluster-level services (e.g. metrics-server) as well
>   as node-level agents (e.g. specific version of CRI). Focus on external or
>   optional services that are needed. For example, if this feature depends on
>   a cloud provider API, or upon an external software-defined storage or network
>   control plane.
> 
>   For each of these, fill in the following—thinking about running existing user workloads
>   and creating new ones, as well as about cluster-level services (e.g. DNS):
>   - [Dependency name]
>     - Usage description:
>       - Impact of its outage on the feature:
>       - Impact of its degraded performance or high-error rates on the feature:
> 

Reviewer 2: ```suggestion
n/a (this KEP pertains to build-time dependency management, run-time dependencies are irrelevant)
```

### Scalability

> _For alpha, this section is encouraged: reviewers should consider these questions
> and attempt to answer them._
> 
> _For beta, this section is required: reviewers must answer these questions._
> 
> _For GA, this section is required: approvers should be able to confirm the
> previous answers based on experience in the field._
> 
> * **Will enabling / using this feature result in any new API calls?**
>   Describe them, providing:
>   - API call type (e.g. PATCH pods)
>   - estimated throughput
>   - originating component(s) (e.g. Kubelet, Feature-X-controller)
>   focusing mostly on:
>   - components listing and/or watching resources they didn't before
>   - API calls that may be triggered by changes of some Kubernetes resources
>     (e.g. update of object X triggers new updates of object Y)
>   - periodic API calls to reconcile state (e.g. periodic fetching state,
>     heartbeats, leader election, etc.)
> 
> * **Will enabling / using this feature result in introducing new API types?**
>   Describe them, providing:
>   - API type
>   - Supported number of objects per cluster
>   - Supported number of objects per namespace (for namespace-scoped objects)
> 
> * **Will enabling / using this feature result in any new calls to the cloud 
> provider?**
> 
> * **Will enabling / using this feature result in increasing size or count of 
> the existing API objects?**
>   Describe them, providing:
>   - API type(s):
>   - Estimated increase in size: (e.g., new annotation of size 32B)
>   - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
> 
> * **Will enabling / using this feature result in increasing time taken by any 
> operations covered by [existing SLIs/SLOs]?**
>   Think about adding additional work or introducing new steps in between
>   (e.g. need to do X to start a container), etc. Please describe the details.
> 
> * **Will enabling / using this feature result in non-negligible increase of 
> resource usage (CPU, RAM, disk, IO, ...) in any components?**
>   Things to keep in mind include: additional in-memory state, additional
>   non-trivial computations, excessive access to disks (including increased log
>   volume), significant amount of data sent and/or received over network, etc.
>   This through this both in small and large cases, again with respect to the
>   [supported limits].

Reviewer 2: ```suggestion
n/a (our existing release-blocking scalability tests do not use bazel-built artifacts)
```

### Troubleshooting

> The Troubleshooting section currently serves the `Playbook` role. We may consider
> splitting it into a dedicated `Playbook` document (potentially with some monitoring
> details). For now, we leave it here.
> 
> _This section must be completed when targeting beta graduation to a release._
> 
> * **How does this feature react if the API server and/or etcd is unavailable?**
> 
> * **What are other known failure modes?**
>   For each of them, fill in the following information by copying the below template:
>   - [Failure mode brief description]
>     - Detection: How can it be detected via metrics? Stated another way:
>       how can an operator troubleshoot without logging into a master or worker node?
>     - Mitigations: What can be done to stop the bleeding, especially for already
>       running user workloads?
>     - Diagnostics: What are the useful log messages and their required logging
>       levels that could help debug the issue?
>       Not required until feature graduated to beta.
>     - Testing: Are there any tests for failure mode? If not, describe why.
> 
> * **What steps should be taken if SLOs are not being met to determine the problem?**
> 
> [supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
> [existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos

Reviewer 2: ```suggestion
n/a (bazel-built artifacts are not used for run-time troubleshooting assistance)
```

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

> - The make system works best from x86 (though it can cross compile all platforms), largely due to a [bug in the kube-cross image build] and some other small oversights / lack of attention, which we should fix anyhow. 

Reviewer 2: ```suggestion
- The make system works best from x86 (though it can cross compile all platforms), largely due to a [bug in the kube-cross image build] and some other small oversights. Improving the existing make system is an explicit non-goal of this KEP, but we support addressing these oversights, and expect the outcome of this KEP will naturally increase available bandwidth to do so.
```

Reviewer 2: /cc @hasheddan 
@dims @hasheddan is there a relevant umbrella issue for non-x86 supported platforms? I know of https://github.com/kubernetes/kubernetes/issues/97550 for development on M1

Reviewer 4: @spiffxp yes, tracking here: https://github.com/kubernetes/sig-release/issues/1337

Author: Support for these platforms is somewhat orthogonal to support for building releases on those platforms (instead of cross compiling). At the very least there will be increased interest in compiling on arm64 though.

  - A few other issues related to CGO / $CC on non-amd64 build hosts have already received pull requests / fixes.
  - Remaining issues largely appear trivial, and should be fixed regardless, especially with the proliferation of non-amd64 developer hardware, where contributors will want to build releases matching the official release (with make).

> - Some contributors may be used to the Bazel CLI, however indications are that this is not true for the majority of contributors. Existing discussions on the matter of removing Bazel from the Kubernetes project have received overwhelming support, most (but certainly not all) of the few suggestions against it do not appear to be from [community members] / active upstream contributors. See: [kubernetes/kubernetes#88533]

Reviewer 3: agree. we are not going to take the long term pain/maintenance just to keep things going.

- In Kubernetes' CI we have bazel [remote caching] enabled which theoretically reduces our resource consumption and improves build times. In practice this gain is reduced currently, due to enabling multiple runs for unit tests to eliminate flakiness. Current measurements show equivalent runtime for the two builds.
  - `go build` has developed high quality caching we can leverage if we need, we have a prototype of this already
  - Kubernetes has enabled repeated runs of "unit" tests to surface flakes, which causes both `go` and Bazel to not cache test results
  - We've disabled caching for large, poorly cachable objects like Docker images anyhow

## Alternatives

<!--
What other approaches did you consider, and why did you rule them out? These do
not need to be as detailed as the proposal, but should include enough
information to express the idea and why it was not acceptable.
-->

> 1. Continue maintaining both build systems

Reviewer 3: non-starter! we tried it, it does not work, we cant keep it going, it's too costly. time to rip the bandaid off!

    - much of the k8s codegen uses fake go build tags, which would require an external process (i.e. gazelle extension) to turn into build files
    - additionally, some codegen relies on certain dependencies, which the gazelle extension would need to figure out
    - many of the code generators are optimized to load the entire Go tree, then generate lots of code at once, which is somewhat incompatible with Bazel's approach to working on a package-by-package level. Generating code package-by-package is much slower (due to having to reparse the tree each time).
  - separate "hack/tools" go module for linters etc: not only do linters not work well (because any source change busts the cache anyhow), but rules_go does not do multiple go modules well. Multiple go modules allows us to isolate development dependencies (like linters) from release binary dependencies, easing dependency management

> 2. Improve bazel integration and drop the make based build

Reviewer 3: -100000

In addition to the points made in 1.) above as to why this is not particularly viable for some of our existing development patterns:

> - We would need to improve support for CGO or eliminate CGO in Kubernetes, both of which would be somewhat expensive to develop

Reviewer 5: Any other technical issue apart from cgo/rules_go integration (and the mentioned below consumed-by-go-get aspect)? It seems like the technical issues left to fix in rules_go/cgo integration are not insurmountable.

If these were to be tracked in an issue somewhere (and referred to here) it would make technical discussion easier.

Author: See other portions of the doc. CGO comments are completely unsupported by rules_go and have no intention of being supported because they don't meet bazel's concepts of "hermetic" (though C/C++ under bazel in general doesn't anyhow).

We depend on hacking around this in brittle ways currently:
https://github.com/kubernetes/kubernetes/blob/81d8ccfa8e641f7295af6dbfe5bbb3c157a10b55/hack/update-bazel.sh#L108-L113

https://github.com/bazelbuild/bazel-gazelle/issues/203#issuecomment-504397957
https://github.com/bazelbuild/bazel-gazelle/blob/987c29003d1616c0b2499fa7f5ce0f3bc11bc9a3/language/go/fileinfo.go#L399

Author: Note that this is not a technical issue to surmount: rules_go has outright rejected support. https://github.com/bazelbuild/bazel-gazelle/pull/556#issuecomment-504397812

  - At minimum kubelet definitely requires CGO for OS integration (e.g. selinux)
  - CGO pkg-config directives do not work in `rules_go`, requiring brittle work-arounds
  - Cross compiling with CGO under bazel is tricky, and despite @ixdy getting close at one point, never shipped in Kubernetes, let alone portably, or capable of shipping a full release.
    - @ixdy suspects this would largely have to be reimplemented now

> - This is unlikely to be popular with contributors

Reviewer 5: That doesn't seem like a real problem? If contributors have technical issues with the Bazel workflow, they should be outlined here explicitly. If these are non-technical (eg. dislike for the sake of dislike), then these probably shouldn't be taken into account when making this decision.

Author: SIG Testing serves to improve the workflow of the project's contributors.
SIG Release exists to ensure we provide high quality binary releases.

Reducing developer friction is absolutely a concern. Our developers are not happy with this system (see previous discussion).

Reviewer 5: Sure, but then these friction issues should IMO be explicitly stated within this KEP. It should obviously show that Bazel-only is a no-go.

Without knowing what the current pain points with Bazel are (and not these with two buildsystems in parallel!), it's impossible to discuss specifics.

Reviewer 2: I would re-frame as: this is less likely to grow the amount of potential contributor bandwidth available to improve the kubernetes/kubernetes build system.  We have a larger pool of contributors today who have demonstrated experience updating our existing make build system vs. updating bazel's components and our usage of it.

Reviewer 2: https://github.com/kubernetes/kubernetes/issues/88553 contains links to specific technical issues we've had in the past with bazel, I don't think it's necessary to re-enumerate here

Author: Thank you @spiffxp, that's the intent of this admittedly poorly written bullet, the "popularity" contest is surely not relevant, but the availability of experience and support within the project is / interest in doing so. I also think https://github.com/kubernetes/kubernetes/issues/88553 has sufficient technical details and is referenced in the doc.

I will update this bullet.

Author: This has been replaced with a variant of @spiffxp's comment.

> - Large portions of Kubernetes are intended for dual use internally and as exported go libraries ("staging") to be consumed by other projects, where we'd still need to support use without bazel (i.e. checked-in generated code etc.).

Reviewer 5: Isn't this already a solved problem even with the current Bazel build? `go build`ing `kuberntes/kubernetes` is atotally different can of worms from `go build`ing re-exported libraries anyway.

Author: > Isn't this already a solved problem even with the current Bazel build? go building kuberntes/kubernetes is atotally different can of worms from go building re-exported libraries anyway.

This means that we don't get to gain from integrating deeply with bazel, because all of this code must already work without it. (so for example code generation must not be bazel style, it needs to be checked in / on disk).

Reviewer 5: I'm pretty sure you could integrate deeply with Bazel here, by continuing to use code-generating rules within `kubernetes/kubernetes`, and then leveraging these same rules to export generated code to `staging/` repositories when publishing.

Author: Regardless we must ensure this code builds for all supported targets with just `go`, it is not feasible to have a world where we only qualify with bazel, unless the entire go / Kubernetes ecosystem adopts bazel, which is unrealistic.

Testing these with bazel has lead to surprises in the past, see the linked issue about moving away from it in kubernetes/kubernetes.

this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

n/a. Not relevant to upgrades. Existing release builds and upgrade CI use make.

### Version Skew Strategy

<!--
If applicable, how will the component handle version skew with other
components? What are the guarantees? Make sure this is in the test plan.

Consider the following in developing a version skew strategy for this
enhancement:
- Does this enhancement involve coordinating behavior in the control plane and
  in the kubelet? How does an n-2 kubelet without this feature available behave
  when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

n/a.

## Production Readiness Review Questionnaire

> **TODO**: This entire section seems completely irrelevant for KEPs that do not

Author: this is TBD, I'm not sure what the policy is here, but I don't think PRR is actually relevant to a KEP like this with literally no change to release artifacts.

Reviewer 6: Discussing KEP types here: https://github.com/kubernetes/enhancements/issues/2311
Drop some notes when you have a chance! :)

Author: I'll take another pass at these fields and the metadata in general tomorrow, and definitely take a look at #2311

Reviewer 2: I suggested edits to drop the text under each heading to "n/a" with explanation as to why, but we can address as followup
